src/jinja2/tatsu.py

```diff
@@ -0,0 +1,137 @@
+import pprint
+
+from tatsu import parse
+
+from .lexer import Lexer
+from .utils import LRUCache
+
+# cache for the lexers. Exists in order to be able to have multiple
+# environments with the same lexer
+_tatsu_lexer_cache = LRUCache(50)
+
+
+def get_tatsu_lexer(environment):
+    """Return a tatsu lexer which is probably cached."""
+    key = (
+        environment.block_start_string,
+        environment.block_end_string,
+        environment.variable_start_string,
+        environment.variable_end_string,
+        environment.comment_start_string,
+        environment.comment_end_string,
+        environment.line_statement_prefix,
+        environment.line_comment_prefix,
+        environment.trim_blocks,
+        environment.lstrip_blocks,
+        environment.newline_sequence,
+        environment.keep_trailing_newline,
+    )
+    tatsu_lexer = _tatsu_lexer_cache.get(key)
+    if tatsu_lexer is None:
+        tatsu_lexer = TatsuLexer(environment)
+        _tatsu_lexer_cache[key] = tatsu_lexer
+    return tatsu_lexer
+
+
+GRAMMAR = r"""
```

Comment:
Nitpick: make this private, i.e. `_GRAMMAR`

Reply:
Updated